tee-apps/sealed-bid-auction/src/core/AuctionService.ts

```diff
@@ -0,0 +1,86 @@
+
+import type {AuctionQuote, AuctionRequest} from "../api/types.ts";
+import type {SolverPriceBook} from "./SolverPriceBook.ts";
+import type {PriceListItem} from "./types.ts";
+
+type Price = {
+    amountOut: bigint;
+    solverAddress: string;
+}
+
+export class AuctionService {
+    constructor(private readonly solverPricebook: SolverPriceBook) {}
+
+    public auction(request: AuctionRequest): AuctionQuote | null {
+        const pricesForAskedTokens = this.findPricesForAskedTokens(request);
+
+        const bestPrice = this.chooseBestPrice(pricesForAskedTokens);
+
+        if (!bestPrice) {
+            throw new Error("No quote found for this pair");
+        } else {
+            return {
+                id: request.id,
+                request: request as Omit<AuctionRequest, "id">,
+                amountOut: bestPrice.amountOut,
+                solverAddress: bestPrice.solverAddress,
+                timestamp: Date.now()
+            };
+        }
+    }
+
+    private getFinalIntervalIndex(priceListItem: PriceListItem, amount: bigint): number | undefined {
+        return priceListItem.intervals.findIndex(interval => interval.range.min <= amount && interval.range.max >= amount);
+    }
+
+    private chooseBestPrice(pricesForAskedTokens: Immutable.List<Price>): Price | null {
+        let bestPrice = null;
+
+        if (!pricesForAskedTokens.isEmpty()) {
+            let bestPrice: Price = {
+                amountOut: -1n,
+                solverAddress: "0xdeadbeef"
+            };
+
+            pricesForAskedTokens.forEach(price => {
+                if (price.amountOut > bestPrice.amountOut) bestPrice = price;
+            });
+        }
+
+        return bestPrice;
+    }
+
+    private findPricesForAskedTokens(request: AuctionRequest): Immutable.List<Price> {
+        return this.solverPricebook.getCurrentPrices().flatMap(
+            priceItems => priceItems.filter(
+                priceItem =>
+                    priceItem.srcTokenAddresses.includes(request.srcTokenAddress) &&
+                    priceItem.dstTokenAddress.includes(request.dstTokenAddress) &&
+                    this.getFinalIntervalIndex(priceItem, request.amountIn) !== undefined
+            ).map(priceItem => {
+                return {
+                    amountOut: this.calculateAmountOut(priceItem, request.amountIn),
+                    solverAddress: priceItem.solverAddress
+                };
+            })
+        )
+    }
+
+    private calculateAmountOut(priceItem: PriceListItem, amountIn: bigint): bigint {
+        let currentInterval = 0;
+        let amountOut = 0n;
+        const finalIndex = this.getFinalIntervalIndex(priceItem, amountIn)!;
+
+        do {
+            const currInterval = priceItem.intervals[currentInterval]!;
+            if (currInterval !== priceItem.intervals[finalIndex]!) {
+                amountOut += currInterval.price * (currInterval.range.max - currInterval.range.min);
```

Comment:
Looks like we’re short‑changing every full interval by one unit.

Why? We treat ranges as *inclusive* on both ends:

* `getFinalIntervalIndex` uses `<=` and `>=`, so an amount equal to `range.max` still belongs to that interval.
* In price‑list validation we insist that `prev.max + 1 === next.min`, which only makes sense if `prev.max` itself is part of the previous interval.
* There’s a little “`+1n`” hack later for the last interval that tries to patch the missing unit.

Given that, an interval `[min, max]` contains `max - min + 1` units, not `max - min`.

```suggestion
// inclusive range: add +1 to capture the max element
const size = currInterval.range.max - currInterval.range.min + 1n;
amountOut += currInterval.price * size;

Reply:
Out of date.